src/v/iceberg/values_json.cc

```diff
@@ -0,0 +1,674 @@
+            throw std::invalid_argument(fmt::format(
+              "Expected offset part for timestamptz_value, got '{}'", str));
+        }
+        auto [u_str, _] = split.value();
+        auto us_sub_sec = sv_to_us(u_str);
+        auto since_epoch = sv_to_time_point("%FT%T", t_str).time_since_epoch();
+        return timestamptz_value{(since_epoch + us_sub_sec) / 1us};
+    }
+    value operator()(const string_type&) {
+        auto str = parse_string_view(data_, "string_value");
+        return string_value{iobuf::from(str)};
+    }
+    value operator()(const fixed_type& t) {
+        auto str = parse_string_view(data_, "fixed_value");
+        if (t.length * 2 != str.size()) {
+            throw std::invalid_argument(fmt::format(
+              "Expected {} hex digits for {}: got {}",
+              t.length * 2,
+              t,
+              str.size()));
+        }
+        return fixed_value{hex_str_to_iobuf(str)};
+    }
+    value operator()(const uuid_type&) {
+        auto str = parse_string_view(data_, "uuid_value");
+        try {
+            return uuid_value{uuid_t::from_string(str)};
+        } catch (const boost::wrapexcept<std::runtime_error>& e) {
+            throw std::invalid_argument(
+              fmt::format("Failed to parse uuid: {}", e.what()));
+        }
+    }
+    value operator()(const binary_type&) {
+        auto str = parse_string_view(data_, "binary_value");
+        return binary_value{hex_str_to_iobuf(str)};
+    }
+    value operator()(const decimal_type& t) {
+        // TODO(oren): need to support negative scale? see datatypes.h
+        auto str = parse_string_view(data_, "decimal_value");
+
+        auto split = split_char(str, '.');
+        if (!split.has_value()) {
+            throw std::invalid_argument(fmt::format(
+              "Expected decimal point for decimal_value, got '{}'", str));
+        }
+
+        auto [int_part, frac_part] = split.value();
+
+        // precision can widen (not narrow) but scale is fixed
+        if (auto json_scale = frac_part.size(); json_scale != t.scale) {
+            throw std::invalid_argument(fmt::format(
+              "Expected {} scale for decimal_value, got {}",
+              t.scale,
+              json_scale));
+        }
+        auto json_precision = int_part.size() + frac_part.size();
+
+        if (json_precision > t.precision) {
+            throw std::invalid_argument(fmt::format(
+              "Expected at most {} precision for decimal_value, got {}",
+              t.precision,
+              json_precision));
+        }
+
+        std::string int_frac;
+        int_frac.reserve(int_part.size() + frac_part.size());
+        std::ranges::copy(int_part, std::back_inserter(int_frac));
+        std::ranges::copy(frac_part, std::back_inserter(int_frac));
+
+        decimal_value result{};
+        if (!absl::SimpleAtoi(int_frac, &result.val)) {
+            throw std::invalid_argument("Failed to parse int128");
+        }
```

Comment:
Generally when parsing decimals I find it easier to do something like: SimpleAtoi the int part, then multiply by scale, SimpleAtoi the fractional part and add it to the previous.

Here's the algorithm I used in Connect: https://github.com/redpanda-data/connect/blob/9d94381914a014c23cb2dd410193e5b7727d4bbd/internal/impl/snowflake/streaming/int128/decimal.go#L146 

That one also rounds up which is standard for IEEE floating point numbers when they are parsed and we have to lose precision.

I also had to support scientific notation, do we need to do that here?


Lastly, I think this algorithm assumes that the fractional part is padded to the scale?

Reply:
>rounds up

Looks like the Java implementation uses BigDecimal, which I believe just gives up if there are too many digits after the point. I'm sort of inclined to truncate the excess and call it a day...

Reply:
Yeah thats the default, but it's configurable. I am fine with truncating. You specified a bad default if it's more precise

Reply:
> configurable

what I mean to say is that the apache code allows it to error :) 

> you specified a bad default if it's more precise

ya. basically UB